runtime/interpreter/interpreter.go

```diff
@@ -1620,150 +1620,104 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
-		}
+		return ConvertUInt256(interpreter, value, locationRange)
 
 	// Word*
 	case sema.Word8Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertWord8(interpreter, value, locationRange)
-		}
-
+		return ConvertWord8(interpreter, value, locationRange)
 	case sema.Word16Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertWord16(interpreter, value, locationRange)
-		}
-
+		return ConvertWord16(interpreter, value, locationRange)
 	case sema.Word32Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertWord32(interpreter, value, locationRange)
-		}
-
+		return ConvertWord32(interpreter, value, locationRange)
 	case sema.Word64Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertWord64(interpreter, value, locationRange)
-		}
+		return ConvertWord64(interpreter, value, locationRange)
 
 	// Fix*
-
 	case sema.Fix64Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertFix64(interpreter, value, locationRange)
-		}
-
+		return ConvertFix64(interpreter, value, locationRange)
 	case sema.UFix64Type:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertUFix64(interpreter, value, locationRange)
-		}
+		return ConvertUFix64(interpreter, value, locationRange)
 	}
 
 	switch unwrappedTargetType := unwrappedTargetType.(type) {
 	case *sema.AddressType:
-		if !valueType.Equal(unwrappedTargetType) {
-			return ConvertAddress(interpreter, value, locationRange)
-		}
+		return ConvertAddress(interpreter, value, locationRange)
 
 	case *sema.ReferenceType:
-		if !valueType.Equal(unwrappedTargetType) {
-			switch ref := value.(type) {
-			case *EphemeralReferenceValue:
-				return NewEphemeralReferenceValue(
-					interpreter,
-					unwrappedTargetType.Authorized,
-					ref.Value,
-					unwrappedTargetType.Type,
-				)
+		switch ref := value.(type) {
+		case *EphemeralReferenceValue:
+			return NewEphemeralReferenceValue(
+				interpreter,
+				unwrappedTargetType.Authorized,
+				ref.Value,
+				unwrappedTargetType.Type,
+			)
 
-			case *StorageReferenceValue:
-				return NewStorageReferenceValue(
-					interpreter,
-					unwrappedTargetType.Authorized,
-					ref.TargetStorageAddress,
-					ref.TargetPath,
-					unwrappedTargetType.Type,
-				)
+		case *StorageReferenceValue:
+			return NewStorageReferenceValue(
+				interpreter,
+				unwrappedTargetType.Authorized,
+				ref.TargetStorageAddress,
+				ref.TargetPath,
+				unwrappedTargetType.Type,
+			)
 
-			case *AccountReferenceValue:
-				return NewAccountReferenceValue(
-					interpreter,
-					ref.Address,
-					ref.Path,
-					unwrappedTargetType.Type,
-				)
+		case *AccountReferenceValue:
+			return NewAccountReferenceValue(
+				interpreter,
+				ref.Address,
+				ref.Path,
+				unwrappedTargetType.Type,
+			)
 
-			default:
-				panic(errors.NewUnexpectedError("unsupported reference value: %T", ref))
-			}
+		default:
+			panic(errors.NewUnexpectedError("unsupported reference value: %T", ref))
 		}
 	}
```

Comment:
Maybe panic with an unexpected error here: The new check at the top filters which types need conversion – if it allowed a type, and we reached this place, the type's conversion isn't implemented properly (feel free to even add this as a comment)

Reply:
While adding this, it made me realize that certain values that were allowed from the above check don't need conversion. e.g: Numeric super types.
Now I'm having second thoughts about this check. Having the checks in two places can easily lead to inconsistencies. 
I now think it maybe good to keep it as is.